docs/core/v3.0.rst

```diff
@@ -0,0 +1,1762 @@
+and values onto file contents. The ``get`` operation could then be
+implemented by reading a file, the ``set`` operation implemented by
+writing a file, and the ``list_dir`` operation implemented by listing
+a directory.
+
+For example, a store implementation might use a key-value database
+such as BerkeleyDB or LMDB as the underlying storage system. In this
+case the implementation of ``get`` and ``set`` operations would be
+whatever native operations are provided by the
+database for getting and setting key/value pairs. Such a store
+implementation might natively support the ``list`` operation but might
+not support ``list_prefix`` or ``list_dir``, although these could be
+implemented via ``list`` with post-processing of the returned keys.
+
+For example, a store implementation might use a cloud object storage
+service such as Amazon S3, Azure Blob Storage, or Google Cloud Storage
+as the underlying storage system, mapping keys to object names and
+values to object contents. The store interface operations would then
+be implemented via concrete operations of the service's REST API,
+i.e., via HTTP requests. E.g., the ``get`` operation could be
+implemented via an HTTP GET request to an object URL, the ``set``
+operation could be implemented via an HTTP PUT request to an object
+URL, and the list operations could be implemented via an HTTP GET
+request to a bucket URL (i.e., listing a bucket).
+
+The examples above are meant to be illustrative only, and other
+implementations are possible. This specification does not attempt to
+standardise any store implementations, however where a store
+implementation is expected to be widely used then it is recommended to
+create a store implementation spec and contribute it to the `zarr-specs GitHub repository`_.
+For an example of a store implementation spec, see the
+:ref:`file-system-store-v1` specification.
+
+
+Storage
+=======
+
+This section describes how to translate high level operations to
+create, erase or modify Zarr hierarchies, groups or arrays, into low
+level operations on the key/value store interface defined above.
+
+In this section a "hierarchy path" is a logical path which identifies
+a group or array node within a Zarr hierarchy, and a "storage key" is
+a key used to store and retrieve data via the store interface. There
+is a further distinction between "metadata keys" which are storage
+keys used to store metadata documents, and "chunk keys" which are
+storage keys used to store encoded chunks.
+
+Note that any non-root hierarchy path will have ancestor paths that
+identify ancestor nodes in the hierarchy. For example, the path
+"/foo/bar" has ancestor paths "/foo" and "/".
+
+.. _storage-keys:
+
+Storage keys
+------------
+
+The entry point metadata document is stored under the key ``zarr.json``.
+
+For a group at a non-root hierarchy path `P`, the metadata key for the
+group metadata document is formed by concatenating "meta/root", `P`,
```

Comment:
As discussed on the community meeting, this naming scheme has the drawback in that there is not a good way to have a path directly to a non-root array.  Additionally, it was noted that to better integrate with existing filesystem completion in editors, etc., it would be helpful if the path were a real filesystem path.

A few proposals were made:

- Instead of `"meta/root" + P + ".group.json"` as the key for the metadata file, instead have it just be: `P + ".zr3"`.  This would lead to a key of just `".zr3"` for the root.  It was noted though that with a consolidated metadata extension these metadata files would not actually exist.  The data could be placed in e.g. "_data" or something, where names beginning with an underscore could be reserved to prevent conflicts.  This would still require a way to locate the root directory --- that could be done by storing the relative path inside the array metadata.
- Alternatively, we could leave the directory structure alone but require an extension as part of the directory containing the root, e.g. "foo.zr3", and disallow any array or group names from ending in ".zr3".  Then you could use: "path/to/root.zr3/path/to/array" as a pseudo-path to an individual array.  The downside is that it may be confusing to use something that looks like a path, and where a portion corresponds to a real filesystem path but a portion does not.  File completion in editors also wouldn't support this.
- We could use a special syntax to combine both a path to a root and a path to an array into a single string, e.g. "path/to/root//path/to/array" or "path/to/root#/path/to/array".  We would need to carefully choose the syntax to avoid conflicts with e.g. fsspec, and file completion in editors also wouldn't support this.

Reply:
Extracted this into an issue and added some more comments I remembered: #177